src/intermediate.rs

```diff
@@ -640,25 +640,147 @@ mod idents {
 }
 pub use idents::*;
 
+/// Details on how to encode a rust type in CBOR. Order is important
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum CBOREncodingOperation {
+    /// CBOR tagged type
+    Tagged(usize),
+    /// bytes .cbor T in cddl, outside of serialization is semantically like T
+    CBORBytes,
+}
+
+/// A complete rust type, including serialization options that don't impact other areas
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct RustType {
+    /// Conceptual type i.e. how it's used in non-serialization contexts
+    pub conceptual_type: ConceptualRustType,
+    /// How to encode the conceptual type. Order is important. Applied in iteration order.
+    pub encodings: Vec<CBOREncodingOperation>,
+    // default value when missing in deserialization
+    pub default: Option<FixedValue>,
+}
+
+impl std::ops::Deref for RustType {
+    type Target = ConceptualRustType;
+
+    fn deref(&self) -> &Self::Target {
+        &self.conceptual_type
+    }
+}
+
+impl RustType {
+    pub fn new(conceptual_type: ConceptualRustType) -> Self {
+        Self {
+            conceptual_type,
+            encodings: Vec::new(),
+            default: None,
+        }
+    }
+
+    pub fn as_alias(mut self, alias_ident: AliasIdent) -> Self {
+        self.conceptual_type = ConceptualRustType::Alias(alias_ident, Box::new(self.conceptual_type));
+        self
+    }
+
+    pub fn tag(mut self, tag: usize) -> Self {
+        self.encodings.push(CBOREncodingOperation::Tagged(tag));
+        self
+    }
+
+    pub fn default(mut self, default_value: FixedValue) -> Self {
+        assert!(self.default.is_none());
+        // TODO: verify that the fixed value makes sense for the conceptual_type
+        self.default = Some(default_value);
+        self
+    }
+
+    pub fn as_bytes(mut self) -> Self {
+        self.encodings.push(CBOREncodingOperation::CBORBytes);
+        self
+    }
+
+    pub fn resolve_aliases(self) -> Self {
+        Self {
+            conceptual_type: self.conceptual_type.resolve_aliases(),
+            encodings: self.encodings,
+            default: self.default,
+        }
+    }
+
+    pub fn cbor_types(&self) -> Vec<CBORType> {
+        match self.encodings.last() {
```

Comment:
why encodings is vec, could probably be an option, I don't see us using any value other than last one.

Reply:
It's a vec because there are cases where it could be multiple things. There are a couple areas this could come up in, like `int` (the primitive case here which calls `Primitive::cbor_types()`) which could be cbor's `uint` or `nint`, or here the structs could be `array` or `map` when they're a plain group (e.g. `foo = (a, b, c)` where the encoding isn't known until it's used somewhere and thus could be used as either but could potentially be known if we add a check to see if it's used only as one ever and never the other but this really isn't a big deal). I can add a comment here explaining why it's a vec. There are parts of the code for implementation details (especially minor things like this function) that aren't the best documented since before September I was the only develop to touch it for the first 2+ years.

Reply:
Oh sorry, you meant `self.encodings`. It's a vector because things are applied in order. e.g. you could have a CBOR encoded tagged type or a tagged CBOR encoded type, which would represent totally different things, which is why I mention in the field comment that the order matters. So in this context for the cbor_types we only care about the outermost CBOR data which if you had the vec as [tag, bytes] the tag wouldn't matter as the whole thing would be encoded into bytes. Or if it were [bytes, tag] it would be encoded as bytes, but then the bytestring would be tagged, so the tag would show up first in CBOR. We use `self.encodings` in many other places where we need to care about the entire thing.